src/VisualStudio/Core/Def/Implementation/DebuggerIntelliSense/DebuggerIntellisenseFilter.cs

```diff
@@ -17,26 +18,70 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.DebuggerIntelliSense
 {
-    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable
+    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable, IFeatureController
         where TPackage : AbstractPackage<TPackage, TLanguageService>
         where TLanguageService : AbstractLanguageService<TPackage, TLanguageService>
     {
         private readonly ICommandHandlerServiceFactory _commandFactory;
-        private readonly IWpfTextView _wpfTextView;
+        private readonly IFeatureServiceFactory _featureServiceFactory;
         private AbstractDebuggerIntelliSenseContext _context;
         private IOleCommandTarget _originalNextCommandFilter;
+        private IFeatureDisableToken _completionDisabledToken;
+        private object _completionDisabledTokenLock = new object();
 
         internal bool Enabled { get; set; }
 
         public DebuggerIntelliSenseFilter(
             AbstractLanguageService<TPackage, TLanguageService> languageService,
             IWpfTextView wpfTextView,
             IVsEditorAdaptersFactoryService adapterFactory,
-            ICommandHandlerServiceFactory commandFactory)
+            ICommandHandlerServiceFactory commandFactory,
+            IFeatureServiceFactory featureServiceFactory)
             : base(languageService, wpfTextView, adapterFactory, commandFactory)
         {
-            _wpfTextView = wpfTextView;
             _commandFactory = commandFactory;
+            _featureServiceFactory = featureServiceFactory;
+        }
+
+        internal void EnableCompletion()
+        {
+            var featureService = _featureServiceFactory.GetOrCreate(WpfTextView);
+            if (!featureService.IsEnabled(PredefinedEditorFeatureNames.Completion))
```

Comment:
If some other caller has _also_ disabled completion through this API, why are we bailing? It seems like the rest of this should still run. Otherwise the EnableCompletion() was ignored; if the other disable is re-enabled, then it'll still not work?

Reply:
1. Please note we are disabling the feature per WpfTextView not globally.

2. If some other caller disabled completion and we do not have a token, what can we do? We can ignore, notify debugger or crash. I discussed sending notifications about "failed to enable" to Dustin from the debugger team. Our API allows to do so. He said they are not interested to be informed about this.

---
In reply to: [270161962](https://github.com/dotnet/roslyn/pull/32631#discussion_r270161962) [](ancestors = 270161962)

Reply:
I imagine the intent of the API to be cooperative: any number of disables can happen, independently. The feature in the end is only active if there aren't _any_ disables. If somebody else has disabled completion too, that's not the concern of this code. Just cancel your disabling, and whatever is going on globally is their problem.

Reply:
I do not think that the feature service API implements the idea you described. They provide only the Disable action (local and global). However, to Enable the feature back you need to know the token. If you do not have the token in hand, you are helpless.

With this restrictions, a better transparent API that we could build should respond to the debugger that we either succeeded to enable or disable or failed to do so. I had this implementation in an intermediate version a couple months ago. However, Dustin said the debugger is not ready to work with it.

I agree the debugger could do all (most of) the work on their side: they can create a token and then dispose it.

---
In reply to: [270216214](https://github.com/dotnet/roslyn/pull/32631#discussion_r270216214) [](ancestors = 270216214)

Reply:
I think the debugger may not know that they call Roslyn (not some other provider) and should disable the feature service for the case. If they should do the same for all providers, they definitely should do this on their side.

---
In reply to: [270228721](https://github.com/dotnet/roslyn/pull/32631#discussion_r270228721) [](ancestors = 270228721,270216214)

Reply:
> They provide only the Disable action (local and global). However, to Enable the feature back you need to know the token. If you do not have the token in hand, you are helpless.

I think the editor API is well designed. But here you should just be enable/disable with a token, and not looking at the effective result. Figuring out the effective result is the completion provider, but shouldn't be done here.

All I'm saying is just delete the IsEnabled call -- you just don't need it here, but it means something else also calling enable/disable can introduce bugs.

Reply:
OK. Thank you! Let us try to remove the check and the call of GetOpenDocumentInCurrentContextWithChanges

---
In reply to: [270232538](https://github.com/dotnet/roslyn/pull/32631#discussion_r270232538) [](ancestors = 270232538)

Reply:
It seems to work. Thank you!

---
In reply to: [271440215](https://github.com/dotnet/roslyn/pull/32631#discussion_r271440215) [](ancestors = 271440215,270232538)